samza-test/src/test/java/org/apache/samza/test/processor/TestZkStreamProcessor.java

```diff
@@ -0,0 +1,558 @@
+      int messageCount) {
+    Map<String, String> configs = new HashMap<>();
+    configs.putAll(StandaloneTestUtils
+            .getStandaloneConfigs("test-job", "org.apache.samza.test.processor.TestZkStreamProcessor.TestStreamTask"));
+    configs.putAll(StandaloneTestUtils
+        .getKafkaSystemConfigs(testSystem, bootstrapServers(), zkConnect(), null, StandaloneTestUtils.SerdeAlias.STRING,
+            true));
+    configs.put("task.inputs", String.format("%s.%s", testSystem, inputTopic));
+    configs.put("app.messageCount", String.valueOf(messageCount));
+    configs.put("app.outputTopic", outputTopic);
+    configs.put("app.outputSystem", testSystem);
+    configs.put(ZkConfig.ZK_CONNECT, zkConnect());
+
+    configs.put("job.systemstreampartition.grouper.factory",
+        "org.apache.samza.container.grouper.stream.GroupByPartitionFactory");
+    configs.put("task.name.grouper.factory", "org.apache.samza.container.grouper.task.GroupByContainerIdsFactory");
+
+    configs.put(JobCoordinatorConfig.JOB_COORDINATOR_FACTORY, "org.apache.samza.zk.ZkJobCoordinatorFactory");
+
+    return configs;
+  }
+
+  /**
+   * Produces the provided number of messages to the topic.
+   */
+  private void produceMessages(final int start, String topic, int numMessages) {
+    KafkaProducer producer = getKafkaProducer();
+    for (int i = start; i < numMessages + start; i++) {
+      try {
+        System.out.println("producing " + i);
+        producer.send(new ProducerRecord(topic, String.valueOf(i).getBytes())).get();
+      } catch (InterruptedException | ExecutionException e) {
+        e.printStackTrace();
+      }
+    }
+  }
+
+  /**
+   * Runs the provided stream processor by starting it, waiting on the provided latch with a timeout,
+   * and then stopping it.
+   */
+  private Thread runInThread(final StreamProcessor processor, CountDownLatch latch) {
+    Thread t = new Thread() {
+
+      @Override
+      public void run() {
+        processor.start();
+        try {
+          // just wait
+          synchronized (this) {
+            this.wait(100000);
+          }
+          System.out.println("notifed. Abandon the wait.");
+        } catch (InterruptedException e) {
+          e.printStackTrace();
+        }
+        System.out.println("Stopping the processor");
+        processor.stop();
+      }
+    };
+    return t;
+  }
+
+  /**
+   * Consumes data from the topic until there are no new messages for a while
+   * and asserts that the number of consumed messages is as expected.
+   */
+  private void verifyNumMessages(String topic, final Map<Integer, Boolean> expectedValues, int expectedNumMessages) {
+    KafkaConsumer consumer = getKafkaConsumer();
+    consumer.subscribe(Collections.singletonList(topic));
+
+    Map<Integer, Boolean> map = new HashMap<>(expectedValues);
+    int count = 0;
+    int emptyPollCount = 0;
+
+    while (count < expectedNumMessages && emptyPollCount < 5) {
```

Comment:
It would be helpful to put a comment why we use 5 as the threshold for checking against emptyPollCount